File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
import heapq

def solution(n, works):
    if n >= sum(works):
        return 0
    
    works = [-work for work in works]
    heapq.heapify(works)
    
    for i in range(n):
        max_work = heapq.heappop(works)
        max_work += 1
        heapq.heappush(works, max_work)
    
    answer = sum(work ** 2 for work in works)
    return answer

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.PriorityQueue;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);

        for (int work : works) {
            pq.add(work);
        }

        while (n > 0 && !pq.isEmpty()) {
            int high = pq.poll();
            if (high == 0) break;
            pq.add(high - 1);
            n--;
        }

        while (!pq.isEmpty()) {
            int tmp = pq.poll();
            answer += (long) tmp * tmp;
        }

        return answer;
    }
}
